File: python/euler/bin/p0050.py
# project euler: problem 50

from collections.abc import Generator
from itertools import dropwhile

from euler.lib.prime import is_prime, prime_generator


def cumsum_generator() -> Generator[int, None, None]:
    acc = 0
    p_gen = prime_generator()
    while True:
        acc += next(p_gen)
        yield acc


# Returns a cumulative sum list of prime numbers.
#   [0, p1, p1+p2, p1+p2+p3, ..., p1+...+p{n-1}, p1+...+p{n-1}+p{n}]
#     where sum(p1..p{n-1}) < limit and sum(p1..p{n}) >= limit
def init_cumsum_lst(cs_gen: Generator[int, None, None], limit: int) -> list[int]:
    lst = [0]
    while lst[-1] < limit:
        lst.append(next(cs_gen))

    return lst


def compute(limit: int) -> str:
    assert limit > 2, 'upper limit must be larger than 2'
    cs_gen = cumsum_generator()
    cs_lst = init_cumsum_lst(cs_gen, limit)

    k = len(cs_lst) - 2
    left = 0
    while True:
        diff = cs_lst[left + k] - cs_lst[left]
        if diff >= limit:
            left = 0
            k -= 1
        elif is_prime(diff):
            return str(diff)
        else:
            left += 1
            if left + k >= len(cs_lst):
                cs_lst.append(next(cs_gen))

    raise RuntimeError('unreachable!')


def solve() -> str:
    return compute(1_000_000)

File: typescript/euler/bin/p0050.ts
// project euler: problem 50

import { isPrime, primeGenerator } from "../lib/prime.ts";

const cumSumGenerator = function* (): Generator<number, void, void> {
  let acc = 0;
  const p_gen = primeGenerator();
  while (true) {
    acc += p_gen.next().value!;
    yield acc;
  }
  // not reached
};

const initCumSumLst = (
  cs_gen: Generator<number, void, void>,
  limit: number,
): number[] => {
  const lst = [0];
  while (lst.at(-1)! < limit) {
    lst.push(cs_gen.next().value!);
  }

  return lst;
};

export const compute = (limit: number): string => {
  if (limit < 3) {
    throw new RangeError("limit must be larger than 2");
  }

  const cs_gen = cumSumGenerator();
  const cs_lst = initCumSumLst(cs_gen, limit);

  let k = cs_lst.length - 2;
  let left = 0;
  while (true) {
    const diff = cs_lst[left + k] - cs_lst[left];
    if (diff >= limit) {
      left = 0;
      k -= 1;
    } else if (isPrime(diff)) {
      return String(diff);
    } else {
      left += 1;
      if (left + k >= cs_lst.length) {
        cs_lst.push(cs_gen.next().value!);
      }
    }
  }
};

export const solve = (): string => compute(1_000_000);
